libsplinter/src/service/scabbard/client/submit.rs

```diff
@@ -0,0 +1,107 @@
+// Copyright 2020 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Contains functions which assist with batch submission to a REST API
+
+use std::{fmt, str};
+
+use protobuf::Message;
+use reqwest::{
+    blocking::{Client, RequestBuilder, Response},
+    Url,
+};
+
+use sawtooth_sdk::messages::batch::BatchList;
+
+use crate::service::scabbard::{BatchInfo, BatchStatus};
+
+use super::Error;
+
+const BATCHES_ENDPOINT: &str = "/batches";
+const HTTP_SCHEME: &str = "http";
+
+pub fn submit_batches(url: &str, batches: BatchList) -> Result<String, Error> {
+    let url = parse_http_url(&format!("{}{}", url, BATCHES_ENDPOINT))?;
+
+    let body = batches.write_to_bytes()?;
+
+    let request = Client::new().post(url).body(body);
+    let response = perform_request(request)?;
+
+    let batch_link: Link = response.json().map_err(|err| {
+        Error::new_with_source("failed to parse response as batch link", err.into())
+    })?;
+
+    Ok(batch_link.link)
+}
+
+pub fn wait_for_batches(url: &str, wait: u64) -> Result<(), Error> {
+    let url = parse_http_url(&format!("{}&wait={}", url, wait))?;
```

Comment:
The wait times should be considered separately here. The wait time this fn takes should be the total time the client wants to wait before returning an error to the user. The wait time in the URL is the wait time for this specific call -- which should be a maximum of 300 seconds to remain below network router-level timeouts (the wait time this fn takes should be unbounded). Thus, this fn should loop until the user-provided wait time is over, calling the URL each iteration of the loop. Also, the server makes no guarantee that it will honor the wait parameter provided in the URL.

Reply:
A story has been captured for this